### PR TITLE
debounce rate input in reimburse expenses

### PR DESCRIPTION
**Pull request by aimane-chnaif**
### Details
Let user keep typing and show raw input value until debounce timeout.
After timeout, fix value and show corrected value accordingly, along with syncing to server
And limit characters length to 12

### Fixed Issues
$ https://github.com/Expensify/App/issues/11352
PROPOSAL: https://github.com/Expensify/App/issues/11352#issuecomment-1271725848

### Tests
1. Login with any account
2. Create workspace if not exist
3. Go to Settings -> workspace -> Reimburse expenses
4. Edit rate
5. Verify that numbers don't shake

- [x] Verify that no errors appear in the JS console

#### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
1. Login with any account
2. Create workspace if not exist
3. Go to Settings -> workspace -> Reimburse expenses
4. Edit rate
5. Verify that numbers don't shake

- [ ] Verify that no errors appear in the JS console

#### Web

https://user-images.githubusercontent.com/96077027/196505882-6ab6e00a-9395-40e0-bdc0-90da3a404e84.mov

#### Mobile Web - Chrome

https://user-images.githubusercontent.com/96077027/196505722-b9b16334-df0b-4c09-9ee4-c5400f068524.mp4

#### Mobile Web - Safari

https://user-images.githubusercontent.com/96077027/196505782-06307935-9292-49b2-80b2-f2df08185aa9.mov

#### Desktop

https://user-images.githubusercontent.com/96077027/196505613-62ec1740-6895-4f3b-96df-4804127b8981.mov

#### iOS

https://user-images.githubusercontent.com/96077027/196505664-6dbaf5c2-b604-4f5a-b39d-118af78aed58.mov

#### Android

https://user-images.githubusercontent.com/96077027/196505554-e99eb571-cc62-4c03-a6e8-642326aff946.mp4

### CODE REVIEW

**Review by thesahindia — CHANGES_REQUESTED** [review]
@aimane-chnaif, I found that the user can add non numeric values to the input. Can we fix that?

**Comment by thesahindia** [comment]
> So on web/desktop (that doesn't show keyboard), use original logic that don't use debounce.
But on native/mWeb (that shows keyboard), use debounce

@aimane-chnaif, I don't like the idea of using two logics just to prevent users from entering non numeric characters. Let me know if you have any other idea.

@AndrewGable, what do you think, do we wanna prevent user from entering non numeric characters? 

After this PR the user will be able to type non numeric chars. If the user refreshes the page or closes the page and comes back then they will see the previous valid value.

**Comment by thesahindia** [comment]
> Thoughts on this?

Let's wait for @AndrewGable's opinion at https://github.com/Expensify/App/pull/11967#issuecomment-1285575911, because I am not 100% sure if we should prevent the user from adding non numeric characters because we also try to not restrict the user from typing something.

**Comment by thesahindia** [comment]
Bump @AndrewGable ^

**Comment by AndrewGable** [comment]
Thanks for the bump, currently we do not allow non-numerical characters to be inserted, so I do not think we should "go back" in a sense. So I think only numbers should allowed to be inserted.

**Comment by thesahindia** [comment]
> value.replace(/[^0-9.]|\.(?=.*\.)/g, '')
Filter only decimal values before applying logic.

Let's do it @aimane-chnaif

**Comment by thesahindia** [comment]
@aimane-chnaif, can you share the screen recording?

**Comment by thesahindia** [comment]
> @thesahindia I updated comment above. So allowing multiple dots (`/[^0-9.]/g`) is my final suggestion

Cool, let's move forward with it.

File: src/pages/workspace/reimburse/WorkspaceReimburseView.js [hunk 1]
```diff
@@ -236,12 +235,13 @@ class WorkspaceReimburseView extends React.Component {
                                     <TextInput
                                         label={this.props.translate('workspace.reimburse.trackDistanceRate')}
                                         placeholder={this.state.outputCurrency}
-                                        onChangeText={value => this.setRate(value)}
+                                        onChangeText={value => this.setRate(value.replace(/[^0-9.]/g, ''))}
```
Comment (thesahindia) [inline src/pages/workspace/reimburse/WorkspaceReimburseView.js:238]: I think we should move this logic inside `setRate` or let's create a new method where we will replace the non numeric chars and call `setRate` inside it with the new value?

**Comment by thesahindia** [comment]
I was able to add 11 digits before the decimal
Screenshot 2022-11-01 at 10 19 44 PM

**Review by thesahindia — CHANGES_REQUESTED** [review]

**Comment by thesahindia** [comment]
Looks good but we have one problem. We are removing `,` even when the language is spanish

https://user-images.githubusercontent.com/83179295/199587599-3d9ba255-a2e6-4711-842c-3ad653989acc.mov

**Comment by thesahindia** [comment]
I think we shouldn't allow user to type `,` or `.` unless they enter a number
Screenshot 2022-11-03 at 4 18 10 PM

**Comment by thesahindia** [comment]
> There's `.`, `,` in mobile keyboard, so if we don't allow at the beginning (remove immediately if typed), shaking issue will happen again on mobile app

Good point. I think the other option is to use the previous valid value after the debounce timeout.

**Comment by thesahindia** [comment]
> This is what I implemented already, isn't it? `this.updatedValue` is the previous valid value to be used after timeout

If I only add `.` or `,` it doesn't update to the previous value.

**Comment by thesahindia** [comment]
> This happens because empty value (after removing dots) is considered as valid rate. As we won't save non-numeric value (including empty) in backend ([reference](https://github.com/Expensify/App/blob/25c306015d7b08fec915e6005cde7c928bdc847e/src/pages/workspace/reimburse/WorkspaceReimburseView.js#L175-L180)), it's safe to remove this condition:

I think it can be a problem, e.g. A user clears the default value and pauses before adding the new value, now the user will have to clear the value again because the previous value will be added again after the timeout.

**Comment by thesahindia** [comment]
<details>
<summary>
Screenshots
</summary>

**Web**




https://user-images.githubusercontent.com/83179295/199950311-f3b0e7a2-ff83-4f9a-8319-2bceacdbd2ac.mov






**Mobile Web - Chrome**





https://user-images.githubusercontent.com/83179295/199950354-b2779cf4-2404-4080-ae90-1822d7d8dd3d.mov





**Mobile Web - Safari**






https://user-images.githubusercontent.com/83179295/199950386-e26e13f9-9ccd-4a08-a29e-6e37cba73546.mov






**Desktop**




https://user-images.githubusercontent.com/83179295/199950398-30bb9a26-e80c-480c-9ac9-8abaead1728b.mov





**iOS**


https://user-images.githubusercontent.com/83179295/199950411-d4bd1a78-a2de-4cbb-afc7-379958013708.mov









**Android**







https://user-images.githubusercontent.com/83179295/199951592-cd610f2f-95ad-4db1-9637-fd2512f675a7.mov




</details>

Looks good and tests well!

cc: @AndrewGable 

C+ reviewed 🎀👀🎀

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Review by thesahindia — APPROVED** [review]

**Comment by thesahindia** [comment]
@aimane-chnaif, thanks for the changes!

**Comment by thesahindia** [comment]
Bump @AndrewGable

**Review by AndrewGable — APPROVED** [review]

**Pull request merged**